### PR TITLE
Add medal processor

### PR DESCRIPTION
**Pull request by peppy**
PRing ahead of adding proper testability (currently looking at updating tests to user a more portable means).

Currently only processes a minimal subset of medals, but is designed to be modular and performant enough to handle all existing medals.

- [x] Depends on #110 
- [x] Depends on #114

### CODE REVIEW

**Comment by peppy** [comment]
I'm still continuing forward work on this, but also believe this PR is in a good state to be reviewed and potentially merged.

File: osu.Server.Queues.ScoreStatisticsProcessor.Tests/MedalProcessorTests.cs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+                setScoreForBeatmap(beatmap.beatmap_id);
+            }
+
+            assertAwarded(medal_id);
+
+            WaitForDatabaseState("SELECT playcount FROM osu_user_stats WHERE user_id = 2", AllBeatmaps.Count, CancellationToken);
+        }
+
+        [Fact]
+        public void TestPlayMultipleBeatmapsFromSameSet()
```
Comment (smoogipoo) [inline osu.Server.Queues.ScoreStatisticsProcessor.Tests/MedalProcessorTests.cs:88]: As discussed IRL, I think these names should be updated to reflect the assertion. For this case maybe something like `TestPlayMultipleBeatmapsFromSameSetDoesNotAward`.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Server.Queues.ScoreStatisticsProcessor.Tests/MedalProcessorTests.cs (hunk 1 above)
Comment (peppy) [reply osu.Server.Queues.ScoreStatisticsProcessor.Tests/MedalProcessorTests.cs]: I've added more documentation and tidied things up. See how it reads now.

File: osu.Server.Queues.ScoreStatisticsProcessor/Processors/IMedalAwarder.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+using osu.Game.Online.API.Requests.Responses;
+using osu.Server.Queues.ScoreStatisticsProcessor.Models;
+
+namespace osu.Server.Queues.ScoreStatisticsProcessor.Processors
+{
+    /// <summary>
+    /// A class which handles conditional awarding of one of more medals.
+    /// </summary>
+    [UsedImplicitly]
+    internal interface IMedalAwarder
```
Comment (smoogipoo) [inline osu.Server.Queues.ScoreStatisticsProcessor/Processors/IMedalAwarder.cs:16]: Probably want this to be `public` for hush-hush.

```suggestion
    public interface IMedalAwarder
```

**Review by smoogipoo — APPROVED** [review]
Seems okay.

**Pull request merged**